feat: CPLYTM-1088 Enforce coverage threshold

sonupreetam opened the pull request:
## Description
- This PR enforces coverage threshold
- Steps Changed: Generate coverage and Enforce Threshold.
- Note: Ignore Sonar scan error as it is running from fork.

## Related Issues
- Refs CPLYTM-1078
- Refs CPLYTM-1090 - Dependent on previously https://github.com/complytime/complybeacon/pull/36
- Use the above workflow and add the enforce scan to the workflow
## Changes
- `.github/workflows/test-coverage.yml`

sonupreetam commented:
@AlexXuan233 Thank you for your feedback. The failing of sonar scan is `normal` here as secrets from root don't propagate to fork repo. The main logic change enforce threshold and coverage report. Also, some additional info, I have added the continue on error for now not to disturb the ongoing workflows but we can remove this future, if the need is explicit.

AlexXuan233 commented:
> @AlexXuan233 Thank you for your feedback. The failing of sonar scan is `normal` here as secrets from root don't propagate to fork repo. The main logic change enforce threshold and coverage report. Also, some additional info, I have added the continue on error for now not to disturb the ongoing workflows but we can remove this future, if the need is explicit.

Hi Sonu. Got it. Thanks for your explanation.

AlexXuan233 reviewed (approved):
LGTM

marcusburghardt commented on .github/workflows/test-coverage.yml:
```diff
@@ -80,17 +93,85 @@ jobs:
           echo "$(go env GOPATH)/bin" >> "$GITHUB_PATH"
           go install github.com/wadey/gocovmerge@latest
-          covs=$(find $modules -maxdepth 1 -type f -name coverage.out -print)
+          # shellcheck disable=SC2086
```
We could safely attend the SC2086 requirement here. I think it would be a good idea actually.

marcusburghardt reviewed (changes requested):
@sonupreetam , I think we could make it simpler. Please, take a look in my comments.

sonupreetam commented on .github/workflows/test-coverage.yml:
```diff
@@ -80,17 +93,85 @@ jobs:
           go install github.com/wadey/gocovmerge@latest
-          covs=$(find $modules -maxdepth 1 -type f -name coverage.out -print)
+          # shellcheck disable=SC2086
+          covs=$(find $modules -maxdepth 1 -type f \
```
@marcusburghardt This is also the same lint warning I found while checking for shell checks.

sonupreetam commented on .github/workflows/test-coverage.yml:
```diff
@@ -80,17 +93,85 @@ jobs:
+
+            # Calculate coverage from merged file (including ALL code)
+            # Format: path:lines num_statements execution_count
+            pct_all=$(awk '
```
@marcusburghardt Also, to explain this clearly, so that not too much effort goes from your end too for this. When we run go test -coverprofile=coverage.out in each module, each coverage file has different import paths and has no native merging. This is why this was custom tailored and also to match SonarCloud's expectations. Let me know if this explains your suggestion.

marcusburghardt replied:
So, your the coverage in the job you mentioned resulted in 41.6% using the complex `awk` commands. We can have exactly the same result using something like this:
```bash
modules=$(git ls-files '**/go.mod' | grep -v '^vendor/' | sed 's#/go\.mod##' | tr '\n' ' ' | sed 's/ $//')
for m in $modules; do
  go test -v -covermode=atomic -coverprofile=coverage.out ./$m/...
  go tool cover -func=coverage.out
done
gocovmerge compass/coverage.out truthbeam/coverage.out proofwatch/coverage.out > coverage.out
go tool cover -func=coverage.out
```
This would be much simpler, smaller and use native tools instead of a custom `awk` command.

marcusburghardt commented on .github/workflows/test-coverage.yml:
```diff
@@ -93,15 +92,18 @@ jobs:
-          if [ -z "$covs" ]; then
+          
+          mapfile -t covs < <(
+            for m in $modules; do
```
In the previous line, I believe that including `"` around `$modules` would already solve the SC2086. If so, I think the simpler command is better.

sonupreetam commented:
@marcusburghardt Also, does the echo statements are something that adds unnecessary lines, or something you would consider to keep? For me, it was mainly for debugging.

sonupreetam commented:
@marcusburghardt Thank you for your feedback and guidance. I have made the code simpler, please do review. 

- However, if tests failed for one of the modules, it won't create a coverage.out, so sticking to dynamic go coverage. For instance in github action, it revealed failed to parse profiles: open compass/coverage.out: no such file or directory. So, it needs to use the dynamic modules and this dynamic collection that only merges files that exist.

- Also, I added missing the tr '\n' ' ' that converts newlines to spaces, as when modules has newlines, GITHUB_OUTPUT gets confused.

- For the GitHub Actions workflow to work properly, I had to ensure go workspace as it was working in local but failing due conflicts with the workspace configuration.

- The successful [run](https://github.com/complytime/complybeacon/actions/runs/19179252415), for testing and debugging I had added the go files, its now deleted.

marcusburghardt reviewed (approved):
LGTM. Thanks @sonupreetam

gvauter reviewed (approved):
LGTM
